### ISSUE TITLE
Outstanding ELF vuln reported several months ago still crashes

### ISSUE THREAD

**Issue by endrift** [visualboyadvance-m/visualboyadvance-m#784]
I made this ELF in January, reported it privately a number of times, and it's still been ignored endlessly. Responsible disclosure periods or anything of the sort expired months ago.

### Expected behavior

VBA-M should not crash when you open a file.

### Actual behavior

VBA-M crashes when running this ELF file I crafted months ago.

### Steps to reproduce the behavior

Open the attached ELF file.

### ROM details

[vbam-vuln.zip](https://github.com/visualboyadvance-m/visualboyadvance-m/files/5611723/vbam-vuln.zip)

### Option details

N/A

### Build details

Operating System (Windows, Mac, Linux (state distribution), etc.): Doesn't matter

If using Linux, specify if using xorg or Wayland: Doesn't matter

Version of code (pre-built binary version, or commit ref, or just "master"): dfa97d8a916fddd4a0fe0b34a07c44abec29d78a, but this issue has existed for a very long time.

**Comment by rkitover** [comment]
We did not purposefully ignore you, it's just that no one has fixed this yet.

Your work on this is very much appreciated.

I will make no claim that this software is secure and people should treat it as such.

To fix these issues will require a thorough refactor of the ELF loader etc., and I just haven't gotten around to it yet, and it's by far not the highest priority.

**Comment by endrift** [comment]
Pretty sure you could fix this with better bounds checking without needing refactoring.

**Comment by rkitover** [comment]
Sure, but there are lots of places that need bounds checking, so it would be changing a lot of code, and the code is horrible and would benefit from a complete rewrite.

**Comment by easyaspi314** [comment]
Unfortunately, @rkitover is right. This code reeks, and needs a rewrite desperately.

I haven't tried that file yet (I'd be more willing to try it if it was a more minimal example), but looking at `elf.cpp`, it seems that the patch for #255

https://github.com/visualboyadvance-m/visualboyadvance-m/blob/bb903e5e08846363c07eb59a7021f4285ec5d372/src/gba/elf.cpp#L2555-L2590

Is missing from this very similar loop:

https://github.com/visualboyadvance-m/visualboyadvance-m/blob/bb903e5e08846363c07eb59a7021f4285ec5d372/src/gba/elf.cpp#L2620-L2641

However, that patch is also bugged: It ignores the error instead of rejecting the file.

As for the other bugs in this file, most of them are reiterations of the `gets` problem: There is no length parameter, so it assumes everything is in bounds.

I could easily cause it to read out of bounds with an invalid ELF: phony offsets, incomplete LEB128, etc.

**Comment by endrift** [comment]
Regardless of the fact that the functionality should be refactored to make it easier to maintain, treating it as low priority and "not guaranteeing safety" of users who routinely download and run unvalidated files from sketchy websites is a _huge_ mistake.

**Comment by easyaspi314** [comment]
Yes. Considering how I have seen those sketchy websites pull stunts like "1834 - Seemingly Normal Game Dump (U)(1.0).**exe**" and the like, if people are falling for that, it is just as easy to do the same with a .elf.

If rewriting the elf parser from scratch is not viable, perhaps offloading it to a more mature library would be better, as those elf files are pretty damn complicated. Then, all you have to worry about is the content.

**Comment by endrift** [comment]
FWIW, mGBA's ELF loading is done using libelf + some basic bounds checks to make sure that the regions are the right size and go where they say they should. It's maybe a dozen lines. That said, I did find some integer overflows while investigating if similar issues existed in VBA (i.e. as I was making this .elf file).

**Comment by easyaspi314** [comment]
libelf would be a good choice. It also has the accompanying libdw which can handle DWARF symbols, so everything that VBA-M currently supports could theoretically be ported to that library. 

By the way, what is different about this .elf file? I am not seeing it.

**Comment by endrift** [comment]
Different compared to what? It's been posted before, but never in a dedicated bug report.

**Comment by easyaspi314** [comment]
I mean what is causing the integer overflow you were talking about? I presumed it had something to do with section sizes (as that is the obvious bug I saw), but nothing seems to be out of range.

**Comment by endrift** [comment]
If I recall correctly this PoC attempts to load into a buffer just before the ROM buffer, which would be rejected, but a carefully crafted input size managed to bypass this via an integer overflow. But it's also been 11 months since I made this file.

**Comment by endrift** [comment]
using `readelf -l` you can see where things start to fall apart:

```
$ readelf -l vuln.elf

Elf file type is EXEC (Executable file)
Entry point 0x8000000
There are 6 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000000 0x03000000 0x03000000 0x000f4 0x02f18 RW  0x10000
  LOAD           0x010000 0x08000000 0x07ffffff 0x42c04 0x42c04 R E 0x10000
  LOAD           0x060000 0x03000000 0x08042c04 0x02940 0x02940 R E 0x10000
  LOAD           0x062f18 0x03002f18 0x08045544 0x00a88 0x00a88 RW  0x10000
  LOAD           0x070000 0x02000000 0x08045fcc 0x01e84 0x01e84 RW  0x10000
  LOAD           0x071e84 0x02001e84 0x08047e50 0x00008 0x00008 RW  0x10000

 Section to Segment mapping:
  Segment Sections...
   00     .bss
   01     .crt0 .init .text .fini .rodata .eh_frame
   02     .iwram
   03     .data .init_array .fini_array
   04     .ewram
   05     .pad
```

Note the second LOAD header.

**Comment by easyaspi314** [comment]
Oh I think I see... That 0x07ffffff looks super fishy.
```
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000000 0x03000000 0x03000000 0x000f4 0x02f18 RW  0x10000
  LOAD           0x010000 0x08000000 0x07ffffff 0x42c04 0x42c04 R E 0x10000
  LOAD           0x060000 0x03000000 0x08042c04 0x02940 0x02940 R E 0x10000
  LOAD           0x062f18 0x03002f18 0x08045544 0x00a88 0x00a88 RW  0x10000
  LOAD           0x070000 0x02000000 0x08045fcc 0x01e84 0x01e84 RW  0x10000
  LOAD           0x071e84 0x02001e84 0x08047e50 0x00008 0x00008 RW  0x10000

 Section to Segment mapping:
  Segment Sections...
   00     .bss 
   01     .crt0 .init .text .fini .rodata .eh_frame 
   02     .iwram 
   03     .data .init_array .fini_array 
   04     .ewram 
   05     .pad 
```

**Comment by easyaspi314** [comment]
Hmm. Even the patch seems to miss this from what I am seeing.

https://github.com/visualboyadvance-m/visualboyadvance-m/blob/bb903e5e08846363c07eb59a7021f4285ec5d372/src/gba/elf.cpp#L2580-L2590

That would underflow to `(unsigned)-1`, but when the size is added, it would still be < `ROM_MAX`, if my calculations are correct.

**Comment by endrift** [comment]
Yeah that sounds right. That leaves a lot of leeway to underrun the buffer. 2^25-1 bytes of leeway to overwrite in `.data` is a _LOT_.

**Comment by easyaspi314** [comment]
Nah, what would happen is that it would either try to write to `rom[-1]` on 32-bit or `rom[0xffffffff]` on 64-bit. 32-bit would cause some sneaky corruption which might only affect it when freed, but 64-bit would be an immediate page fault.

**Comment by endrift** [comment]
It's only 1 byte because this is a PoC showing it exists at all. That code snippet shows the range checks would pass for values up to 32 MiB - 1, even though I only exercise it as 1 in this.

**Comment by rkitover** [comment]
I'm in favor of a better libelf loader, and your arguments about ROMs from fishy websites is valid, but I can't promise to write one myself, and for the past few months I'm the only dev on this project not counting hit-and-run PRs.

On top of that I have serious health issues right now so I have very little time and energy to work on this compared to what I had before.

So my top priority right now is keeping things running and trying to get another release out.

If you guys want to work on this stuff, I'll be more than happy to help in any way I can.

**Comment by easyaspi314** [comment]
I am going to see if there is a better option than libelf. 

Libelf relies on the compiler being GCC and won't even compile on Clang without modification (it uses nested functions, which clang can sort of emulate with blocks, which are also nonstandard). Considering how it seems that the default Windows build uses MSVC, that definitely won't work.

**Comment by easyaspi314** [comment]
Everything but the CLI utilities (useless to us), AR archive support (useless to us), and file I/O (useless to us, it uses POSIX files, not stdio) seems to compile fine under MSVC when I patch a few missing macros, remove references to the BSD headers, and add the missing `sys/queue.h` header.

This is definitely doable, and would probably be useful for mGBA as well, since DWARF support is intact.

**Comment by rkitover** [comment]
Do you have a repo for your work, I can help with making a vcpkg port for it etc..

**Comment by endrift** [comment]
I've been using libelf on Windows for years, though built with GCC. I'm pretty sure it can output libraries that are linkable with MSVC, but I'm not 100% sure.

**Comment by easyaspi314** [comment]
True. However, the elfutils version can't be built with Clang, which is the "gcc" for macOS, iOS (irrelevant), and the Android NDK.

Side note:
```c
    int res = fread(elfFileData, 1, size, f);
    fclose(f);

    if (res < 0) {
        free(elfFileData);
        elfFileData = NULL;
        return false;
    }
```
```c
size_t fread ( void * ptr, size_t size, size_t count, FILE * stream );
```
I think someone got `read` and `fread` mixed up. :thinking:

**Comment by endrift** [comment]
What do you mean? That's the right invocation, modulo swapping the size and count arguments. `read` doesn't have a count argument anyway.

**Comment by easyaspi314** [comment]
I mean the return value check. fread returns < count, not < 0.

**Comment by andyvand** [comment]
It works now. GBA ELF support is fixed.
